rest: Workaround Http2 zero-length reply bug

MichaelEischer opened the pull request:
What does this PR change? What problem does it solve?
-----------------------------------------------------
The golang http client does not return an error when a HTTP2 reply includes a non-zero content length but does not return any data at all. This scenario can occur e.g. when using rclone when a file stored in a backend seems to be accessible but then fails to download.

This can cause errors like "Pack ID does not match, want $HASH, got e3b0c44" as reported in #2742.

The underlying issue is a bug in the go http client, which is also tracked as https://github.com/golang/go/issues/46071 . However, it is unknown when a fix will be available and I don't expect it to be backported to all go versions still supported by restic.  I've added a test that will fail once the underlying issue in the http.client has been fixed, to allow us to eventually remove the workaround again.

Was the change previously discussed in an issue or on the forum?
----------------------------------------------------------------
Fixes #2742 .
See also https://forum.restic.net/t/http2-stream-closed-connection-reset-context-canceled/3743/10 .

Checklist
---------
- [x] I have read the [contribution guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches).
- [x] I have [enabled maintainer edits](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [x] I have added tests for all code changes.
- ~~[ ] I have added documentation for relevant changes (in the manual).~~
- [x] There's a new file in `changelog/unreleased/` that describes the changes for our users (see [template](https://github.com/restic/restic/blob/master/changelog/TEMPLATE)).
- [x] I have run `gofmt` on the code in all commits.
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits).
- [ ] I'm done! This pull request is ready for review.

MichaelEischer committed:
rest: workaround for HTTP2 zero-length replies bug

The golang http client does not return an error when a HTTP2 reply
includes a non-zero content length but does not return any data at all.
This scenario can occur e.g. when using rclone when a file stored in a
backend seems to be accessible but then fails to download.

MichaelEischer committed:
rest: test that zero-length replies over HTTP2 work correctly

The first test function ensures that the workaround works as expected.
And the second test function is intended to fail as soon as the issue
has been fixed in golang to allow us to eventually remove the
workaround.

MichaelEischer committed:
Add changelog

fd0 commented on internal/backend/rest/rest.go:
```diff
@@ -246,6 +247,20 @@ func (b *Backend) openReader(ctx context.Context, h restic.Handle, length int, o
+	// see also https://forum.restic.net/t/http2-stream-closed-connection-reset-context-canceled/3743/10
+	if resp.ContentLength == 0 && resp.ProtoMajor == 2 && resp.ProtoMinor == 0 {
+		if clens := resp.Header["Content-Length"]; len(clens) == 1 {
+			if cl, err := strconv.ParseUint(clens[0], 10, 63); err == nil {
```
How about returning an error to the caller if the value of the `Content-Length` header cannot be parsed? If it is present but the value is invalid, that's an error. Otherwise, users will just see the "unexpected EOF" error when the server returns garbage in the header.

MichaelEischer replied:
If the golang client is unable to parse `Content-Length`, then `resp.ContentLength` is probably equal to `-1`. And in that case our error handling will be dead code.

Based on the same argument, it should probably be safe to use `resp.Header.Get("Content-Length")`

MichaelEischer replied:
And to immediately contradict myself: Obviously the content-length was not parsed yet, because otherwise it wouldn't be zero...

fd0 commented on internal/backend/rest/rest.go:
```diff
@@ -246,6 +247,20 @@ func (b *Backend) openReader(ctx context.Context, h restic.Handle, length int, o
+			if cl, err := strconv.ParseUint(clens[0], 10, 63); err == nil {
+				resp.ContentLength = int64(cl)
+			}
+			if resp.ContentLength != 0 {
```
I find this error handling very confusing, mainly because we reuse `resp.ContentLength`.

MichaelEischer replied:
Indeed, whenever `ContentLength` gets changed then it's an error...

fd0 committed:
rest: Rework handling HTTP2 zero-length replies bug

Add comment that the check is based on the stdlib HTTP2 client. Refactor
the checks into a function. Return an error if the value in the
Content-Length header cannot be parsed.

fd0 reviewed (commented):
I've left a few comments and I've tried to make handling this case clearer. I've pushed a commit, please have a look!

MichaelEischer commented on internal/backend/rest/rest.go:
```diff
@@ -246,6 +247,20 @@ func (b *Backend) openReader(ctx context.Context, h restic.Handle, length int, o
+	// workaround https://github.com/golang/go/issues/46071
+	// see also https://forum.restic.net/t/http2-stream-closed-connection-reset-context-canceled/3743/10
+	if resp.ContentLength == 0 && resp.ProtoMajor == 2 && resp.ProtoMinor == 0 {
+		if clens := resp.Header["Content-Length"]; len(clens) == 1 {
```
Looks like I've lost the information that the content-length parsing snippet is essentially the code from https://github.com/golang/go/blob/b7a85e0003cedb1b48a1fd3ae5b746ec6330102e/src/net/http/h2_bundle.go#L8646 . My intention was to behave the same way as the golang http2 implementation. As such checking that only a single content-length was returned by the server would ensure that the behavior matches.

fd0 replied:
Ok, I've added the check and a comment to the stdlib code.

fd0 reviewed (approved):
I'm satisfied, please have a look again @MichaelEischer!

MichaelEischer commented:
The cleanup looks fine now.
